app/src/org/commcare/views/widgets/MediaWidget.java

```diff
@@ -162,4 +180,117 @@ public void setFocus(Context context) {
+                Log.e(TAG, "IOExeception while copying media");
+                e.printStackTrace();
+            }
+        }
+
+        if (newMedia.exists()) {
+            // Add the copy to the content provider
+            ContentValues values = new ContentValues(6);
+            values.put(MediaStore.Audio.Media.TITLE, newMedia.getName());
+            values.put(MediaStore.Audio.Media.DISPLAY_NAME, newMedia.getName());
+            values.put(MediaStore.Audio.Media.DATE_ADDED, System.currentTimeMillis());
+            values.put(MediaStore.Audio.Media.DATA, newMedia.getAbsolutePath());
+
+            Uri mediaUri =
+                    getContext().getContentResolver().insert(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, values);
+            String mediaUriString = mediaUri == null ? "null" : mediaUri.toString();
+            Log.i(TAG, "Inserting media returned uri = " + mediaUriString);
+            showToast("form.attachment.success");
+        } else {
+            Log.e(TAG, "Inserting media file FAILED");
+        }
+
+        mBinaryName = newMedia.getName();
+    }
+
+    /**
+     * If file is chosen by user, the file selection intent will return an URI
+     * If file is auto-selected after recording_fragment, then the recordingfragment will provide a string file path
+     * Set value of customFileTag if the file is a recent recording from the RecordingFragment
+     */
+    protected String createFilePath(Object binaryuri) {
+        String path = "";
+        destMediaPath = "";
+        if (binaryuri instanceof Uri) {
+            try {
+                path = UriToFilePath.getPathFromUri(CommCareApplication.instance(),
+                        (Uri)binaryuri);
+            } catch (UriToFilePath.NoDataColumnForUriException e) {
+                // Need to make a copy of file using uri, so might as well copy to final destination path directly
+                InputStream inputStream;
+                try {
+                    inputStream = getContext().getContentResolver().openInputStream((Uri)binaryuri);
+                } catch (FileNotFoundException e1) {
+                    showToast("form.attachment.notfound");
+                    e1.printStackTrace();
+                    return "";
+                }
+
+                recordedFileName = FileUtil.getFileName(((Uri)binaryuri).getPath());
+                destMediaPath = mInstanceFolder + System.currentTimeMillis() + FileUtil.getExtension(((Uri)binaryuri).getPath());
+
+                try {
+                    FileUtil.copyFile(inputStream, new File(destMediaPath));
+                } catch (IOException e1) {
+                    e1.printStackTrace();
+                    showToast("form.attachment.copy.fail");
+                    return "";
+                }
+                path = destMediaPath;
+            }
+            customFileTag = "";
+        } else {
+            path = (String)binaryuri;
+            customFileTag = CUSTOM_TAG;
```

Comment:
What purpose is `customFileTag` serving?

Reply:
I guess we are trying to make a distinction between files recorded vs files chosen from file provider. Though don't really think if it's useful and can remove if it's not serving any pupose right now.